Queue clearance for emergency stop

lromor opened the pull request.

hzeller commented on src/motion-queue.h:
```diff
@@ -118,6 +118,10 @@ class MotionQueue {
+
+  // Perform an immediate reset of the queue,
+  // even if motors are still moving.
+  virtual bool Clear() = 0;
```
The 'even if motors are still moving': does that mean best-effort to stop motors ASAP ? Then this should be mentioned here (Imagine someone implementing it: they read the comment and need to come up with an implementation that matches that expectation).
What does the return value mean ?

hzeller commented on src/motion-queue-motor-operations_test.cc:
```diff
@@ -202,6 +211,35 @@ TEST(RealtimePosition, zero_loops_edge) {
+  // Enqueue a segment
+  LinearSegmentSteps segment = {
+    0 /* v0 */,
+    0 /* v1 */,
```
If both speeds are zero, how would we move ?

lromor replied:
(solved)

hzeller commented on src/pru-motion-queue.cc:
```diff
@@ -177,6 +177,11 @@ void PRUMotionQueue::Shutdown(bool flush_queue) {
 }
 
+bool PRUMotionQueue::Clear() {
+  pru_interface_->Shutdown();
```
this sounds a bit drastic. Is that all we can do ?

hzeller reviewed (commented):
first round of comments.

hzeller commented on src/segment-queue.h:
```diff
@@ -77,6 +77,12 @@ class SegmentQueue {
+  // Clear the queue and reset it to its initial state.
+  // This action is immediate and will discard any running or
+  // enqueued and not yet executed segment.
+  // Current physical status will be lost.
```
What does the return value mean ?

lromor replied:
(solved)

hzeller commented on src/motion-queue-motor-operations_test.cc:
```diff
@@ -202,6 +205,32 @@ TEST(RealtimePosition, zero_loops_edge) {
+  PhysicalStatus status;
+  motor_operations.GetPhysicalStatus(&status);
+  EXPECT_THAT(expected, ::testing::ContainerEq(status.pos_steps));
+  motor_operations.HaltAndDiscard();
```
To keep the readers' train of thought, I'd group the  things with newlines to better see the before/call/after situation:

```c++
  PhysicalStatus status;
  motor_operations.GetPhysicalStatus(&status);
  EXPECT_THAT(expected, ::testing::ContainerEq(status.pos_steps));

  motor_operations.HaltAndDiscard();
  EXPECT_EQ(motion_backend.clear_calls_count, 1);

  memset(expected, 0, sizeof(expected));
  motor_operations.GetPhysicalStatus(&status);
  EXPECT_THAT(expected, ::testing::ContainerEq(status.pos_steps));
```

lromor replied:
(solved)

hzeller reviewed (commented):
LGTM.

Did you verify that stoping and starting will work on the PRU, and that you can restart and start enquing things again that come out on the other oscilloscope end ?

lromor committed:
Queue clearance for emergency stop

Adds the primitives needed to stop and reset the PRU motion queue
without going through the normal shutdown/restart cycle:

  MotionQueue::HaltAndDiscard()          - immediately stop motion,
                                            clear the ring buffer, and
                                            reset the PRU's status.
  PruHardwareInterface::Halt() / Restart() - pause/resume PRU program
                                             execution without losing
                                             the loaded firmware.
  SegmentQueue::HaltAndDiscard()         - wraps the above at the
                                            planner-facing layer, also
                                            clears the shadow history
                                            queue so subsequent status
                                            reports don't reference
                                            discarded motion.

These are the building blocks the ESTOP fix for #65 needs to bring the
machine to a clean, restartable state without waiting for the queue to
drain. Callers wiring HaltAndDiscard into an ESTOP path will land in
a follow-up.
